x/uibc/quota/ibc_module.go

```diff
@@ -0,0 +1,177 @@
+)
+
+type IBCMiddleware struct {
+	app    porttypes.IBCModule
+	keeper keeper.Keeper
+}
+
+// NewIBCMiddleware creates a new IBCMiddlware given the keeper and underlying application
+func NewIBCMiddleware(app porttypes.IBCModule, k keeper.Keeper) IBCMiddleware {
+	return IBCMiddleware{
+		app:    app,
+		keeper: k,
+	}
+}
+
+// OnChanOpenInit implements types.Middleware
+func (im IBCMiddleware) OnChanOpenInit(ctx sdk.Context, order channeltypes.Order, connectionHops []string,
+	portID string, channelID string, channelCap *capabilitytypes.Capability, counterparty channeltypes.Counterparty,
+	version string,
+) (string, error) {
+	return im.app.OnChanOpenInit(
+		ctx,
+		order,
+		connectionHops,
+		portID,
+		channelID,
+		channelCap,
+		counterparty,
+		version,
+	)
+}
+
+// OnChanOpenTry implements types.Middleware
+func (im IBCMiddleware) OnChanOpenTry(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID string, channelID string,
+	channelCap *capabilitytypes.Capability, counterparty channeltypes.Counterparty,
+	counterpartyVersion string) (version string, err error) {
+	return im.app.OnChanOpenTry(
+		ctx, order, connectionHops, portID, channelID, channelCap,
+		counterparty, counterpartyVersion,
+	)
+}
+
+// OnChanOpenAck implements types.Middleware
+func (im IBCMiddleware) OnChanOpenAck(ctx sdk.Context, portID string, channelID string, counterpartyChannelID string,
+	counterpartyVersion string,
+) error {
+	return im.app.OnChanOpenAck(ctx, portID, channelID, counterpartyChannelID, counterpartyVersion)
+}
+
+// OnChanOpenConfirm implements types.Middleware
+func (im IBCMiddleware) OnChanOpenConfirm(ctx sdk.Context, portID string, channelID string) error {
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements types.Middleware
+func (im IBCMiddleware) OnChanCloseInit(ctx sdk.Context, portID string, channelID string) error {
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements types.Middleware
+func (im IBCMiddleware) OnChanCloseConfirm(ctx sdk.Context, portID string, channelID string) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements types.Middleware
+func (im IBCMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	// if this returns an Acknowledgement that isn't successful, all state changes are discarded
+	return im.app.OnRecvPacket(ctx, packet, relayer)
+}
+
+// OnAcknowledgementPacket implements types.Middleware
+func (im IBCMiddleware) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	var ack channeltypes.Acknowledgement
+	if err := json.Unmarshal(acknowledgement, &ack); err != nil {
+		return sdkerrors.ErrUnknownRequest.Wrapf("cannot unmarshal ICS-20 transfer packet acknowledgement: %v", err)
+	}
```

Comment:
shouldn't we call `app.OnAcknowledgementPacket` as well? Otherwise, the app won't be called, and maybe the status will be broken. What do you think?

Reply:
We are calling `app.OnAcknowledgementPacket` end of the function , if marshal failed means some is not coming correctly from middlewares

Reply:
Added TODO to code to recheck the `app.OnAcknowledgementPacket` error handling